doc/decisions/transformations.md

```diff
@@ -0,0 +1,98 @@
+Those transformations include, but are not limited to:
+
+- Changing the names of the keys inbetween `kdbGet` and `kdbSet` operations
+- Changing values back and forth for normalization, e.g. `true` -> `1`, `1` -> `true`
+- Adding and removing metadata (e.g. the `spec` plugin)
+
+While these features are useful, they do create feature-interaction problems.
+More specifically, problems have been observed in conjunction with the following (overlapping) types of plugins:
+
+- notification plugins
+- plugins that do change tracking
+
+The problem, in general, can be described as: Which version of the KDB should be used for notifications/change tracking?
+
+### Observed problems with changing key names
+
+For example, the `rename` plugin supports different transformations for `get` and `set`.
+The plugin is executed in the post-storage phase for the `get` operation and the pre-storage phase for the `set` operation.
+We can instruct it to convert the keynames to lowercase in the `get` operation, and to UPPERCASE in the `set` operation.
+This results in the keys being in UPPERCASE in the configuration files, but they are presented in lowercase to other plugins and applications using the Elektra API.
+
+For example, here is a configuration file with a hypothetical format:
+
+```
+/DISPLAY/BRIGHTNESS = 100
+/DISPLAY/COLOR = #FFFFFF
+``` 
+
+As can be seen, the keys are in UPPERCASE within the configuration file.
+In Elektra keys are case-sensitive. 
+As operations on keysets such as `ksLookup` operate after the post-storage phase, `kdb get /DISPLAY/BRIGHTNESS` will fail.
+For Elektra, the key `/DISPLAY/BRIGHTNESS` does not exist, as the `rename` plugin transformed this into the lowercase `/display/brightness`.
+
+This leads to problems with the notification plugins.
+As notification plugins are executed after the post-storage phase of the `set` operation, they will receive a keyset with the already transformed keys.
+In this example, the notification plugins will receive all-UPPERCASE keys, and send out notifications with those all-UPPERCASE keys. 
+An application listening to those notifications will not be able to query Elektra for those keys, as for Elektra those UPPERCASE keys do not exist.
+
+Apart from the problems with notifications, the way the `rename` plugin works also breaks change tracking in plugins like `dbus`.
+This is because the `rename` plugin does the transformation by deleting the original key and adding a new key with the transformed name.
+The `dbus` plugin implements change tracking by checking the 'key needs sync' flag instead of comparing the values.
+As new keys by design have the 'key needs sync' flag set, the plugins that implement change tracking via the flag will always erroneously detect transformed keys as changed.
+
+### Observed problems with normalization
+
+TBD ... 
+
+In theory the same issue with the 'key needs sync' flag based change tracking approach as above.
```

Comment:
See above: don't center the problem around "key needs sync" flag. This is an implementation detail. This specific problem could be easily solved within the rename plugin but it still wouldn't solve the whole problem.

Maybe we can even drop the whole idea of "key needs sync" and let the change tracking algorithm decide about if changes happened. This is actually a pure performance topic to be answered by a benchmark.

Reply:
Using the change tracking to detect changes should definitely be investigated. I'm not sure how feasible it is, but could be worth a try. It should definitely be more reliable, because it is not an extra flag that needs to be considered and could be forgotten somewhere.

Reply:
Yes, another problem we currently have is that irrelevant changes also cause changes to be triggered, e.g. assuming the key `k` has the content `orig`:

```c
keySetString(k, "abc"); // actually not even needed to trigger the problem
keySetString(k, "orig"); // triggers `k` to need sync
ksAppend(ks, k);
kdbSet(kdb, ks, p); // would write keys even though nothing has changed!
```

Expected would be that `kdbSet` is a NOP as no changes occurred.

Reply:
> Expected would be that kdbSet is a NOP as no changes occurred.

Yes, good observation! I'll add replacing `keyNeedsSync` flag with change tracking as an alternative.